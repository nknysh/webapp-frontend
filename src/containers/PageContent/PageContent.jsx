import React from 'react';
import { Helmet } from 'react-helmet';
import { compose, isEmpty } from 'ramda';
import { useTranslation } from 'react-i18next';

import { AsyncNotFound } from 'pages/NotFound';
import { useLoading } from 'effects';
import { Markdown, Sidebar, Loader } from 'components';

import { propTypes, defaultProps } from './PageContent.props';
import connect from './PageContent.state';
import {
  StyledPageContent,
  Columns,
  ColumnLeft,
  ColumnRight,
  PageContentData,
  PageContainer,
  PageHero,
} from './PageContent.styles';

const renderNotFound = () => <AsyncNotFound />;

export const PageContent = ({ pageId, data, links, title, getPage, className, hero }) => {
  const { t } = useTranslation();

  const loadPage = () => getPage(pageId);
  const loading = useLoading(loadPage, [pageId], true);

  const isLoading = !data && loading;

  if (!data && !loading) {
    return renderNotFound();
  }

  return (
    <Loader isLoading={isLoading}>
      <StyledPageContent className={className}>
        <Helmet>
          <title>
            {title || ''} - {t('title')}
          </title>
        </Helmet>
        {hero && <PageHero {...hero} />}
        <PageContainer>
          <Columns>
            <ColumnLeft>
              <Sidebar title={title}>{!isEmpty(links) && <Markdown>{links}</Markdown>}</Sidebar>
            </ColumnLeft>
            <ColumnRight>
              <PageContentData>
                <Markdown>{data}</Markdown>
              </PageContentData>
            </ColumnRight>
          </Columns>
        </PageContainer>
      </StyledPageContent>
    </Loader>
  );
};

PageContent.propTypes = propTypes;
PageContent.defaultProps = defaultProps;

export default compose(connect)(PageContent);
