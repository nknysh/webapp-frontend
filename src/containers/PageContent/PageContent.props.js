import PropTypes from 'prop-types';

export const propTypes = {
  page: PropTypes.any,
  pageId: PropTypes.string,
  getPage: PropTypes.func,
};

export const defaultProps = {
  page: undefined,
  pageId: undefined,
  getPage: () => {},
};
