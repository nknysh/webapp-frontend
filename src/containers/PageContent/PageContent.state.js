import { connect } from 'react-redux';
import { pipe } from 'ramda';

import { getPageId, getPageTitle, getPageData, getPageLinks, getPageHero } from 'store/modules/pages/selectors';
import { getPage } from 'store/modules/pages/actions';

export const mapStateToProps = state => ({
  data: getPageData(state),
  hero: getPageHero(state),
  id: getPageId(state),
  links: getPageLinks(state),
  title: getPageTitle(state),
});

export const mapDispatchToProps = dispatch => ({
  getPage: pipe(
    getPage,
    dispatch
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
);
