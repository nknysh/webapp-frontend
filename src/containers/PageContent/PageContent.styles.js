import styled from 'styled-components';

import theme from 'styles/theme';
import { Container } from 'styles/elements';
import { Heading1 } from 'styles/typography';

import { Hero } from 'components/Hero';

const fiveSpaced = theme.gutter * 5;
const doubleSpaced = theme.gutter * 2;

export const StyledPageContent = styled.div`
  color: ${theme.colors['gold-neutral']};
  width: 100%;
`;

export const PageHero = styled(Hero)`
  margin-bottom: ${fiveSpaced}px;
`;

export const PageContainer = styled(Container)``;

export const PageContentHeader = styled(Heading1)`
  margin: 0;
  padding: 0 0 ${theme.gutter * 3}px;
  border-bottom: 1px solid ${theme.colors['gray-medium']};
`;

export const PageContentLinks = styled.div`
  ul {
    list-style: none;
    padding: 0;
    margin: 0;

    li {
      padding: ${doubleSpaced}px 0;
      color: ${theme.primary};
      border-bottom: 1px solid ${theme.colors['gray-medium']};

      &:last-child {
        border-bottom: 0;
      }

      a {
        color: ${theme.primary};
        font-size: 11px;
        text-transform: uppercase;
      }
    }
  }
`;

/**
 * @todo Temporary format some internal markup from markdown.
 *       Look how to control styling in a more modular way.
 */
export const PageContentData = styled.div`
  hr {
    border: 0;
    background: 0;
    outline: 0;
    margin: 0;
    padding: 0;
    border-bottom: 1px solid ${theme.colors.gray};
  }

  h2,
  h3 {
    color: ${theme.primary};
  }

  h2 {
    margin: 0 0 ${doubleSpaced}px;
    padding: 0 0 ${doubleSpaced}px;
    border-bottom: 1px solid ${theme.colors['gray-medium']};
    min-height: 35px;
  }

  .list-grid {
    ul {
      display: flex;
      flex-wrap: wrap;
      list-style: none;
      margin: 0;
      padding: 0;

      li {
        flex: 1 1 50%;
        margin: 0 0 ${theme.gutter * 4}px;
        padding: 0;

        h4 {
          color: ${theme.primary};
        }
      }
    }
  }
`;

export const Columns = styled.div`
  display: flex;
`;

export const ColumnLeft = styled.div`
  flex: 1 1 25%;
  padding-right: ${theme.gutter * 15}px;
`;

export const ColumnRight = styled.div`
  flex: 1 1 50%;
  padding-bottom: ${fiveSpaced}px;
`;
