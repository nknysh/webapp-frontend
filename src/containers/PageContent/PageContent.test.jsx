import React from 'react';
import { PageContent } from './PageContent';

const mockGetPage = jest.fn();
const getComponent = props => shallow(<PageContent getPage={mockGetPage} {...props} />);

describe('<PageContent />', () => {
  describe('render', () => {
    it('matches snapshot while loading', () => {
      expect(getComponent()).toMatchSnapshot();
    });
    it('matches snapshot with markdown', () => {
      expect(getComponent({ data: '### some data' })).toMatchSnapshot();
      expect(getComponent({ title: 'foo', data: '### some data' })).toMatchSnapshot();
      expect(getComponent({ title: 'foo', links: '- a link', data: '### some data' })).toMatchSnapshot();
      expect(
        getComponent({ title: 'foo', links: '- a link', data: '### some data', hero: { image: 'http://foo' } })
      ).toMatchSnapshot();
    });
  });
});
