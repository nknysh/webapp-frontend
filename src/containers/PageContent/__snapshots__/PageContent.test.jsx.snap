// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageContent /> render matches snapshot while loading 1`] = `
<Loader
  isLoading={true}
  showSpinner={true}
>
  <PageContentstyles__StyledPageContent>
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
         - 
        Pure Escapes
      </title>
    </HelmetWrapper>
    <PageContentstyles__PageContainer>
      <PageContentstyles__Columns>
        <PageContentstyles__ColumnLeft>
          <Sidebar
            isOpen={false}
            links={Array []}
          >
            <Markdown />
          </Sidebar>
        </PageContentstyles__ColumnLeft>
        <PageContentstyles__ColumnRight>
          <PageContentstyles__PageContentData>
            <Markdown />
          </PageContentstyles__PageContentData>
        </PageContentstyles__ColumnRight>
      </PageContentstyles__Columns>
    </PageContentstyles__PageContainer>
  </PageContentstyles__StyledPageContent>
</Loader>
`;

exports[`<PageContent /> render matches snapshot with markdown 1`] = `
<Loader
  isLoading={false}
  showSpinner={true}
>
  <PageContentstyles__StyledPageContent>
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
         - 
        Pure Escapes
      </title>
    </HelmetWrapper>
    <PageContentstyles__PageContainer>
      <PageContentstyles__Columns>
        <PageContentstyles__ColumnLeft>
          <Sidebar
            isOpen={false}
            links={Array []}
          >
            <Markdown />
          </Sidebar>
        </PageContentstyles__ColumnLeft>
        <PageContentstyles__ColumnRight>
          <PageContentstyles__PageContentData>
            <Markdown>
              ### some data
            </Markdown>
          </PageContentstyles__PageContentData>
        </PageContentstyles__ColumnRight>
      </PageContentstyles__Columns>
    </PageContentstyles__PageContainer>
  </PageContentstyles__StyledPageContent>
</Loader>
`;

exports[`<PageContent /> render matches snapshot with markdown 2`] = `
<Loader
  isLoading={false}
  showSpinner={true}
>
  <PageContentstyles__StyledPageContent>
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        foo
         - 
        Pure Escapes
      </title>
    </HelmetWrapper>
    <PageContentstyles__PageContainer>
      <PageContentstyles__Columns>
        <PageContentstyles__ColumnLeft>
          <Sidebar
            isOpen={false}
            links={Array []}
            title="foo"
          >
            <Markdown />
          </Sidebar>
        </PageContentstyles__ColumnLeft>
        <PageContentstyles__ColumnRight>
          <PageContentstyles__PageContentData>
            <Markdown>
              ### some data
            </Markdown>
          </PageContentstyles__PageContentData>
        </PageContentstyles__ColumnRight>
      </PageContentstyles__Columns>
    </PageContentstyles__PageContainer>
  </PageContentstyles__StyledPageContent>
</Loader>
`;

exports[`<PageContent /> render matches snapshot with markdown 3`] = `
<Loader
  isLoading={false}
  showSpinner={true}
>
  <PageContentstyles__StyledPageContent>
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        foo
         - 
        Pure Escapes
      </title>
    </HelmetWrapper>
    <PageContentstyles__PageContainer>
      <PageContentstyles__Columns>
        <PageContentstyles__ColumnLeft>
          <Sidebar
            isOpen={false}
            links={Array []}
            title="foo"
          >
            <Markdown>
              - a link
            </Markdown>
          </Sidebar>
        </PageContentstyles__ColumnLeft>
        <PageContentstyles__ColumnRight>
          <PageContentstyles__PageContentData>
            <Markdown>
              ### some data
            </Markdown>
          </PageContentstyles__PageContentData>
        </PageContentstyles__ColumnRight>
      </PageContentstyles__Columns>
    </PageContentstyles__PageContainer>
  </PageContentstyles__StyledPageContent>
</Loader>
`;

exports[`<PageContent /> render matches snapshot with markdown 4`] = `
<Loader
  isLoading={false}
  showSpinner={true}
>
  <PageContentstyles__StyledPageContent>
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
    >
      <title>
        foo
         - 
        Pure Escapes
      </title>
    </HelmetWrapper>
    <PageContentstyles__PageHero
      image="http://foo"
    />
    <PageContentstyles__PageContainer>
      <PageContentstyles__Columns>
        <PageContentstyles__ColumnLeft>
          <Sidebar
            isOpen={false}
            links={Array []}
            title="foo"
          >
            <Markdown>
              - a link
            </Markdown>
          </Sidebar>
        </PageContentstyles__ColumnLeft>
        <PageContentstyles__ColumnRight>
          <PageContentstyles__PageContentData>
            <Markdown>
              ### some data
            </Markdown>
          </PageContentstyles__PageContentData>
        </PageContentstyles__ColumnRight>
      </PageContentstyles__Columns>
    </PageContentstyles__PageContainer>
  </PageContentstyles__StyledPageContent>
</Loader>
`;
